src/client/Client.cc

```diff
@@ -7239,17 +7241,16 @@ int Client::readdir_r_cb(dir_result_t *d, add_dirent_cb_t cb, void *p)
   if (dirp->offset == 1) {
     ldout(cct, 15) << " including .." << dendl;
     uint64_t next_off = 2;
-    if (!diri->dn_set.empty()) {
+    if (diri->client->is_root(diri)) {
+      /* must be at the root (no parent), so fill out with same info as "." */
+      fill_stat(diri, &st);
+      fill_dirent(&de, "..", S_IFDIR, st.st_ino, next_off);
+    } else {
       InodeRef& in = diri->get_first_parent()->inode;
```

Comment:
I think you're missing a check for an unlinked directory here. (i.e. dn_set.empty() is true)